design_docs/oscode/oscode_sundials.md

```diff
@@ -0,0 +1,347 @@
+
+$$
+\dot{\mathbf{y}} = \begin{pmatrix}
+\mathbf{y}[1]\\
+-\mathbf{y}[0]\omega^2(t)-2\gamma(t)\mathbf{y}[1]
+\end{pmatrix}
+$$
+
+Among the SUNDIALS modules, ARKODE ---which solves equations with known explicit and implicit, stiff and nonstill timescale components--- is structured most similarly to OSCODE.
+
+$$
+M(t)\dot{\mathbf{y}} = f^E(t, \mathbf{y}) + f^I(t, \mathbf{y}), \quad \mathbf{y}(t_0) = \mathbf{y}_0
+$$
+
+For the new OSCODE module we propose the right-hand-side will be a subset of what ARKODE can handle, specifically,
+
+$$
+M(t)\dot{\mathbf{y}} = f^O(t, \mathbf{y}), \quad \mathbf{y}|(t_0) = \mathbf{y}_0, \\
+$$
+where $f^O(t)$ takes the specific form
+$$
+f^O(t, y) = \begin{pmatrix}
+\mathbf{y}[1]\\
+-\mathbf{y}[0]\omega^2(t)-2\gamma(t)\mathbf{y}[1]
+\end{pmatrix}.
+$$
+
+As the solution $\mathbf{y}(t)$ may vary between oscillatory and smoothly varying, OSCODE will dynamically switch between a Runge-Kutta solver and its asymptotic ARDC-based solver, at each step choosing the method which yields the larger step length (while keeping the local error within user defined tolerance).
+
+
+## Guide-level Explanation
+
+------------------------------------
+Section Goals:
+
+Explain the proposal as if it was already included in the project and you were teaching it to another Sundials programmer in the manual. That generally means:
+
+- Introducing new named concepts.
+- Explaining the feature largely in terms of examples.
+- Explaining how Sundials programmers should *think* about the feature, and how it should impact the way they use the relevant package. It should explain the impact as concretely as possible.
+- If applicable, provide sample error messages, deprecation warnings, or migration guidance.
+
+<!--- Should the above be deleted? --->
+------------------------------------
+
+(Note: Much of this is taken from the ARKODE docs and modified for OSCODE)
+
+<!--- I removed/modified many of the statements because they didn't apply for OSCODE.--->
+
+The OSCODE infrastructure provides adaptive-step time integration modules for a class of nonstiff Ordinary Differential Equations (ODEs) where the components may be highly oscillatory. Current users of ARKODE will notice a similar structure in the C API of OSCODE.
+
+OSCODE supports ODE systems posed in the form
+
+$$
+M(t)\dot{\mathbf{y}} = f^O(t, \mathbf{y}), \quad \mathbf{y}|(t_0) = \mathbf{y}_0, \\
+$$
+where $f^O(t)$ takes the specific form
+$$
+f^O(t, y) = \begin{pmatrix}
+\mathbf{y}[1]\\
+-\mathbf{y}[0]\omega^2(t)-2\gamma(t)\mathbf{y}[1]
+\end{pmatrix}.
+$$
+
+Here, $t$ is the independent variable, $\mathbf{y}$ is a set of two (complex-valued) dependent variables with one being the derivative of the other, $\omega$ and $\gamma$ are user specified, real-valued callables. The solver has the ability to switch dynamically between using a Runge-Kutta method and the adaptive Riccati defect correction method to advance the solution, while adaptively updating its stepsize.
+
```

Comment:
The ARDC method approximates the solution with an asymptotic expansion computed numerically, adaptive in the number of terms it retains in the expansion. The approximation is accurate for many wavelengths of an oscillatory solution provided that $\omega$ is large and $\dot{\omega}/\omega$ is small over the interval, thus allowing OSCODE to increase its stepsize while traversing an oscillatory region. These criteria ensure that there exists a nonoscillatory ``phase function'', an underlying slowly changing quantity, which may be approximated well via Chebyshev (or other) polynomials. The ARDC method provides a local error estimate over its timestep, and is capable of interpolating a highly oscillatory solution at little additional cost (around 1 microsecond per output point) by performing interpolation of the phase function.